services/datamanager/builtin/sync_test.go

```diff
@@ -250,65 +249,76 @@ func TestDataCaptureUpload(t *testing.T) {
+				for i := 0; i < numFails; i++ {
+					mockClock.Add(syncInterval)
+					// Each time we sync, we should get a sync request for each file.
+					for j := 0; j < numFiles; j++ {
+						wait := time.After(time.Second * 5)
+						select {
+						case <-wait:
+							t.Fatalf("timed out waiting for sync request")
+						case <-mockClient.failedDCRequests:
+						}
+					}
+				}
+			}
+
+			mockClient.fail.Store(false)
 			// If testing manual sync, call sync. Call it multiple times to ensure concurrent calls are safe.
 			if tc.manualSync {
 				err = newDMSvc.Sync(context.Background(), nil)
 				test.That(t, err, test.ShouldBeNil)
 				err = newDMSvc.Sync(context.Background(), nil)
 				test.That(t, err, test.ShouldBeNil)
 			}
-			time.Sleep(syncTime)
-			err = newDMSvc.Close(context.Background())
-			test.That(t, err, test.ShouldBeNil)
-
-			// If we set failAt, we want to restart the service to simulate when the backend fails.
-			var newMockService *mockDataSyncServiceServer
-			if tc.serviceFailAt > 0 {
-				// Now turn back on with only sync enabled.
-				test.That(t, rpcServer.Stop(), test.ShouldBeNil)
-				rpcServer, newMockService = buildAndStartLocalSyncServer(t, 0, 0)
-				defer func() {
-					err := rpcServer.Stop()
-					test.That(t, err, test.ShouldBeNil)
-				}()
-
-				newDMSvc.Close(context.Background())
-				newestDMSvc := newTestDataManager(t)
-				defer newestDMSvc.Close(context.Background())
-				newestDMSvc.SetSyncerConstructor(getTestSyncerConstructor(rpcServer))
-				newestDMSvc.SetWaitAfterLastModifiedMillis(testLastModifiedMillis)
-				err = newestDMSvc.Update(context.Background(), cfg)
-				test.That(t, err, test.ShouldBeNil)
-				time.Sleep(syncTime)
-				err = newestDMSvc.Close(context.Background())
-				test.That(t, err, test.ShouldBeNil)
-			}
 
-			// Calculate combined successful/failed requests
-			var successfulURs []*v1.DataCaptureUploadRequest
-			if tc.serviceFailAt > 0 {
-				successfulURs = append(mockService.getSuccessfulDCUploadRequests(), newMockService.getSuccessfulDCUploadRequests()...)
-			} else {
-				successfulURs = mockService.getSuccessfulDCUploadRequests()
+			var successfulReqs []*v1.DataCaptureUploadRequest
+			// Get the successful requests
+			mockClock.Add(syncInterval)
+			for i := 0; i < numFiles; i++ {
+				wait := time.After(time.Second * 5)
+				select {
+				case <-wait:
+					t.Fatalf("timed out waiting for sync request")
+				case r := <-mockClient.succesfulDCRequests:
+					successfulReqs = append(successfulReqs, r)
+				}
 			}
-			failedURs := mockService.getFailedDCUploadRequests()
 
-			// If the server was supposed to fail for some requests, verify that it did.
-			if tc.numFails != 0 || tc.serviceFailAt != 0 {
-				test.That(t, len(failedURs), test.ShouldBeGreaterThan, 0)
+			// Give it time to delete files after upload.
+			waitUntilNoCaptureFiles := func(captureDir string) {
+				files := getAllFileInfos(captureDir)
+				for i := 0; i < 100; i++ {
```

Comment:
[nit] Similarly here and wherever we retry while waiting for files, useful to extract into const at top of file.